`Forget` marker trait (#3782)
rust-lang/rfcs: text/0000-forget-marker-trait.md

[`local_default_bounds`]: https://github.com/rust-lang/rfcs/pull/3783

# Summary
[summary]: #summary

Add a `Forget` marker trait indicating whether it is safe to skip the destructor before the value of a type exits the scope and basic utilities to work with `!Forget` types. Introduce a seamless migration route for the standard library and ecosystem.

# Motivation
[motivation]: #motivation

Many readers may find the biggest problem with `Forget` to be migration. RFC's confidence is taken from the fact that migration can be done easily. See [#migration](#migration) section for details.

Back in 2015, the [decision was made][safe-mem-forget] to make `mem::forget` safe, making every type implicitly implement `Forget`. All APIs in `std` could've been preserved after that change, except one. Today is 2025 and some things changed, old reasoning is no longer true. This RFC is not targeted at resource leaks in general but is instead focused on allowing a number of APIs to become safe by providing new unsafe guarantees.

[safe-mem-forget]: https://github.com/rust-lang/rfcs/pull/1066

### What is a proxy RAII guard?
[proxy-raii-guards]: #proxy-raii-guards

> `thread::scoped` is special because it uses the RAII guard [^raii] as a proxy to represent other values, but this proxy is not used to access those values. Instead, we are trusting that the borrow checker will ensure that the guard cannot outlive those values, and therefore that joining the thread in the guard's destructor is enough to ensure that the spawned thread is no longer running. [^proxy-raii-guard-source]

Author: I probably should refine this section a bit more, to be more clear

    let fut = something_with_clean_up(async |foo| {
        foo.bar().await;
    });
    {
        // Pin the future.
        let pinned = Box::pin(fut);
        // Poll the future once.
        poll_fn(|cx| Poll::Ready(_ = pinned.poll(cx))).await;
        forget(pinned); // or `_ = Box::leak(pinned);`
    }
    // rest of the code...
}
```

The library is only taking control flow in between `await` points. Here, future is pinned and `Pin`'s [drop guarantee] is met (boxed future remains allocated for `'static`), but cleanup cannot run - `Drop` handler of `fut` is skipped. Thus, APIs that require any cleanup for safety can be expressed in `sync` Rust, but not in `async` Rust, making `async` less attractive, as the operating system APIs and C/C++ libraries *cannot* be used efficiently, ergonomically, and safely.

[drop guarantee]: https://doc.rust-lang.org/std/pin/#drop-guarantee

Another important observation that we can make is that `Pin`'s drop guarantee only applies to the memory of the `Future` itself. But if `Future` borrows a buffer, that buffer *can* be deallocated or re-used before the `drop` of the `Future` is called. See [#connection-to-pin](#connection-to-pin).

> ## Examples of unsafe async APIs that can be allowed in sync Rust

Reviewer: You have plenty of examples here already but in case you want one more: WASI 0.3 has an `uring`-like design, and the lack of guaranteed destructors means that for the Rust bindings we have to choose between different options, none of which are great. More context on that here: https://github.com/WebAssembly/component-model/issues/471

Author: Thanks!

[`ringbahn`]: https://github.com/ringbahn/ringbahn/
[`tokio_uring`]: https://docs.rs/tokio-uring/latest/tokio_uring/
[`rio`]: https://lib.rs/crates/rio/

### `take_mut`

The async version of [`take_mut`] cannot be created as it relies on cleanup code to abort the program.

[`take_mut`]: https://docs.rs/take_mut/latest/take_mut/

### C/C++ bindings + async do not work well together
[example-async-c-cpp-bindings]: #example-async-c-cpp-bindings

It is common for C/C++ APIs to require some cleanup. It is not an issue for `sync` rust, as wrappers can just take a closure/callback and ensure that cleanup. But all `async` calls are transformed into `impl Future + use<'a>`, not passing control flow to the wrapper. `io_uring` and `async-cuda` fall into that category too. For embedded/kernel development this issue is even worse, as you often cannot afford an allocation due to the lack of resources or complex locking, making borrows your only option and making `Pin`'s drop guarantee not useful for you.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> The core goal of `Forget` trait, as proposed in that RFC, is to bring back the "Proxy Guard" idiom for non-static types, with `async` being the primary motivation.

Reviewer 2: Do I understand correctly that all `!Forget` types need to have a lifetime bound to have any effect? Does it ever make sense to have a `!Forget` type that has no lifetime? (like an integer)

Reviewer 3: That’s correct.

Author: From the perspective of use cases and definitions, it really makes little sense to have `!Forget` and `'static` together. But I am not opposed too, it may have some other benefits which I am not aware of / or implementational / migrational benefits.

Reviewer 2: This makes me wonder whether `PhantomNotForget` should have a generic lifetime argument. Although it's not technically necessary, it would make it clearer that it's about lifetimes. I could work as self-documenting code which lifetime is the tricky one, and `PhantomNotForget<'static>` would be something to lint against.

Reviewer 3: If there were to be a blanket `impl<T: 'static> Forget for T`, then the lifetime parameter would be necessary for the type to work at all. I think that alone justifies it. IIUC the lifetime parameter would have to be covariant

Author: I would mention it just in case, this is already in unresolved questions:
>  Maybe force impl Forget for T where T: 'static {} and add a generic to the PhantomNonForget? Use cases and unsafe guarantee are fine with it, and we already allow !Forget in static.

Reviewer 4: > This makes me wonder whether `PhantomNotForget` should have a generic lifetime argument. Although it's not technically necessary, it would make it clearer that it's about lifetimes. I could work as self-documenting code which lifetime is the tricky one, and `PhantomNotForget<'static>` would be something to lint against.

In my (not new) reference implementation of the `Forget` trait, I have added it on a wrapper type `Unforget` as a contravariant lifetime:

https://zetanumbers.github.io/leak-playground/leak_playground_std/marker/struct.Unforget.html

As to the magic `'static` lifetime, seems like it made a lot of contention in its prior discussion:

[#t-lang > The destruction guarantee and linear types formulation](https://rust-lang.zulipchat.com/#narrow/channel/213817-t-lang/topic/The.20destruction.20guarantee.20and.20linear.20types.20formulation)

## API changes
[library-api-changes]: #library-api-changes

- `Rc`/`Arc` - all APIs for construction,  except the new `Rc::new_unchecked` method, only exist for `T: Forget` types. If we decide to not have `impl<T: 'static> Forget for T {}`, in the future we *may* allow safe constructors for `T: ?Forget + 'static` (resources are borrowed for `'static`, it fulfills the guarantee we are giving to the unsafe code) and something along the lines of `T: ?Forget + Freeze` (to forbid cycles), author of the RFC is not familiar enough with interior mutability questions.
- `ManuallyDrop<T>` always implements `Forget`, regardless of the `T`. `ManuallyDrop::new` is available for types with `T: Forget`.  New unsafe method `ManuallyDrop::new_unchecked`, available for `T: ?Forget`, is introduced. We may add a safe constructor with `T: ?Forget + 'static`, as we allow forgetting in statics.
- `Box::<T>::into_ptr` is available only for `T: Forget`. As for `T: !Forget` users should `ManuallyDrop::new_unchecked` and take the pointer via `&raw mut`. It will still be allowed to pass this pointer to `Box::from_ptr`.

> - `Box::<T>::forget` is available only for `T: Forget`.

Reviewer 5: ```suggestion
- `Box::<T>::leak` is available only for `T: Forget`.
```

- `forget_unchecked`, a new unsafe function, is added to forget `T: ?Forget` types. It is a wrapper around `ManuallyDrop::new_unchecked`, just as `forget` is a wrapper around `ManuallyDrop::new`.
- `PhantomNonForget` is a `!Forget` ZST for types to become `!Forget`. If we decide to have `impl<T: 'static> Forget for T {}`, we should add a generic parameter/lifetime to `PhantomNonForget`.
- Similar to other APIs, `T: Forget` -> `Drait<T>: Forget` and `T: !Forget` -> `Drait<T>: !Forget`.
- APIs like `std::sync::mpsc::Sender::send` are available only for `T: Forget`.
- Possibly new channels should be introduced, that are compatible with `T: !Forget` types too.
- `Vec::set_len` is available only for `T: Forget` types, to not create a footgun for `unsafe` code in the wild. Maybe a new method should be added to support `T: ?Forget`.
- `ptr::write` is available for all types.

> - Etc

Author: That RFC, currently, does not has a 100% answer, because I personally don't care which of 2 variants to pursue, but some decision makers have more strong opinions, so it should be resolved in a discussion than by text imho.

Options are:
1. Have `impl<T: 'static> impl Forget for T`. This will not allow any `'static + !Forget` to exist.
2. Not have it. But it doesn't exclude aborting, your example, and some others. Unsafe guarantee we are giving does not contradict that example *at all*, because resources borrowed by `T: 'static` are borrowed for `'static`. We guarantee that drop handle will run before that runtime ends - if lifetime never ends, `'static`, you can drop it immediately and walk out satisfied.

Thus, spawn should not be marked `Forget` in either case - maybe in the first, but it is anyway implied.

> - Etc

Author: > that's the reason that this would be rejected?

More precisely, unsafe code authors are not able to write unsound code that would UB in that example, if they use unsafe guarantee language is giving them correctly.

> - Etc

Author: I think it was worth it and I added a notice with a short explanation directly to the PR message.

If the crate is maintained, however, migration should not be difficult.

#### `#![forbid(unsafe)]` crates
[safe-local-defaults-migration]: #safe-local-defaults-migration

1. Set the appropriate bounds:

```rust
// can be with `cfg_attr`
#![default_generic_bounds(?Forget)]
```

2. Resolve any compilation errors by explicitly adding `+ Forget` where needed.

3. Optionally: Recurse into your dependencies, applying the same changes as needed. Most probably you will use `!Forget` types with well-maintained crates providing combinators or containers.

#### For crates with `unsafe` code (like `libcore`)
[unsafe-local-defaults-migration]: #unsafe-local-defaults-migration

1. Set the appropriate bounds:

```rust
#![default_generic_bounds(?Forget)]
```

> 2. Audite your codebase to work properly with `!Forget` types.

Reviewer: ```suggestion
2. Audit your codebase to work properly with `!Forget` types.
```

## Leakpocalypse
[leakpocalypse-prior-art]: #leakpocalypse-prior-art

- https://github.com/rust-lang/rfcs/pull/1066
- https://github.com/rust-lang/rust/issues/24292
- https://cglab.ca/~abeinges/blah/everyone-poops/
- https://github.com/rust-lang/rfcs/pull/1085
- https://doc.rust-lang.org/std/thread/fn.scope.html
- https://smallcultfollowing.com/babysteps/blog/2015/04/29/on-reference-counting-and-leaks/

## Usage of the pattern
[usage-prior-art]: #usage-prior-art

- https://doc.rust-lang.org/std/thread/struct.Builder.html#method.spawn_unchecked
- https://docs.rs/async-task/latest/async_task/fn.spawn_unchecked.html
- https://blog.japaric.io/safe-dma/
- https://docs.rs/async_nursery/latest/async_nursery/
- https://without.boats/blog/the-scoped-task-trilemma/
- https://docs.rs/async-scoped/latest/async_scoped/struct.Scope.html#method.scope_and_collect

## Miscellaneous
[misc-prior-art]: #misc-prior-art

- https://github.com/rust-lang/rfcs/issues/1111
- https://tmandry.gitlab.io/blog/posts/2023-03-01-scoped-tasks/

## MustMove types
[must-move-prior-art]: #must-move-prior-art

- https://faultlore.com/blah/linear-rust/

> - https://blog.yoshuawuyts.com/linear-types-one-pager/

Author: Yes, sure

- https://smallcultfollowing.com/babysteps/blog/2023/03/16/must-move-types/

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> - [ ] Maybe the name `Forget` is misleading, as its core is around the `unsafe` guarantee of borrowed resources and the destructor. `IndirectPin`?

Reviewer 2: I'd rather not mix it with `Pin`. `Pin` is an advanced concept needed only for certain niches, so for many users it may not clarify the meaning, but rather connect it to something they're unfamiliar with. There are also proposals for integrating `Pin` more into Rust's syntax, so the `Pin` itself may become even less visible in the future.

I assume that `Forget` may appear more often in the API surface. It may eventually be easier to teach `Pin` in terms of `!Forget` than the other way.

- [ ] Maybe force `impl Forget for T where T: 'static {}` and add a generic to the `PhantomNonForget`? Use cases and unsafe guarantee are fine with it, and we already allow `!Forget` in `static`.
- [ ] Maybe add `StaticForget<T: ?Forget + 'static>: Forget`.
- [ ] How does it interact with `&own`?

> - [ ] Maybe make `Vec::set_len` available for `T: ?Forget`, but with a new unsafe precondition. Crates with unsafe code that are manually migrating to support `!Forget` would need to be aware of that change and verify/modify their unsafe code to work correctly with `!Forget` types, or manually restrain them to `Forget`.

Reviewer 2: `Vec::set_len` has two distinct purposes depending on whether the length is increased (init) or shrunk (forget).

Maybe it should be split into `Vec::leaking_truncate` and `Vec::extend_from_capacity_assume_init`? The length-extending function doesn't affect drops.

Author: I think this is out of the RFCs scope and is for a separate ACP later 😄

- Feature Name: `forget_marker_trait`
- Start Date: (fill me in with today's date, YYYY-MM-DD)
- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

<!-- todo: Replace with RFC PR later -->
[`local_default_bounds`]: https://github.com/rust-lang/rfcs/pull/3783

# Summary
[summary]: #summary

Add a `Forget` marker trait indicating whether it is safe to skip the destructor before the value of a type exits the scope and basic utilities to work with `!Forget` types. Introduce a seamless migration route for the standard library and ecosystem.

# Motivation
[motivation]: #motivation

Many readers may find the biggest problem with `Forget` to be migration. RFC's confidence is taken from the fact that migration can be done easily. See [#migration](#migration) section for details.

In short, the lack of `!Forget` types undermines lifetimes, sacrificing all performance, ergonomics and efficiency. Rust, as well as external APIs, naturally converge towards `!Forget` types, but old decisions force us to shift to `Arc`, `'static`, allocations, etc.

---

> Back in 2015, the [decision was made][safe-mem-forget] to make `mem::forget` safe, making every type implicitly implement `Forget`. All APIs in `std` could've been preserved after that change, except one. Today is 2025 and some things changed, old reasoning is no longer true. This RFC is not targeted at resource leaks in general but is instead focused on allowing a number of APIs to become safe by providing new unsafe guarantees.

Reviewer 6: It's unusual to say that something that didn't exist yet was implicitly done. I'm not sure what you meant by the middle two sentences, but I've tried to make it clearer.

```suggestion
Back in 2015, the [decision was made][safe-mem-forget] to make `mem::forget` safe, making every type effectively implement `Forget`. By making `mem::forget` safe, other `std` APIs had to become unsafe, or their safe implementations became complex (such as async Rust). This RFC is not targeted at resource leaks in general, but is instead focused on allowing a number of APIs to become safe, by providing new unsafe guarantees using `Forget`.
```

[safe-mem-forget]: https://github.com/rust-lang/rfcs/pull/1066

> ### What is a proxy RAII guard?

Reviewer 6: Before this explanation, it would be helpful to explain why a proxy RAII guard is relevant to `Forget`.

Author: Can I explain why proxy RAII guard is relevant without explaining what is proxy RAII guard first?

Reviewer 6: Imagine that someone is reading the document like a story, so each paragraph or section builds on and connects to the one before.

One way to phrase that is:
> Forget has (short summary of relationship) to proxy RAII guards. For more on proxy RAII guards, see (later section link).
> (Explain in more detail why proxy RAII guards are relevant to Forget.)

Thus, there was no point in redesigning the language and delaying Rust 1.0, practically all APIs and patterns could be safely expressed without destructors always running, so making `std::mem::forget` safe and removing "Proxy RAII Guard" was a good decision at the time.

## What is different
[what-is-different]: #what-is-different

Edition 2018 introduced `async` Rust. But as turned out, nuances in its design conflicted with an earlier decision. All `async` calls are essentially constructors for state machines which borrow some resources from outside or directly own them. It is user's responsibility to poll those state machines to completion.

`!Forget` use cases could've been expressed by other means in sync Rust (like taking a callback instead of returning a guard or [PPYP]), but with `async`, anything turns directly into `impl Future + use<'a>` which is equivalent to the RAII guard. This means, that sync pattern of taking a closure cannot be used - everything is transformed into RAII guard by the compiler.

> Various OS or C/C++ APIs cannot be made `async` without performance or ergonomics costs, because futures become a proxy RAII guard for those APIs. PPYP can work for `Drain<'a>`, but not for `io_uring`. As long as the future is `'static` or directly owns all data it is accessing, `Pin` guarantees are sufficient. Otherwise, there is no way to make a sound API. Currently, they are forced into using `'static` bounds, which is one of the pain points users are reporting about `async` Rust, together with `Send` issues.

Reviewer 6: Please explain acronyms like PPYP the first time you use them. Usually this is done by expanding the acronym in brackets:
> PPYP (Please Provide Your Portrayal)

Author: Pre poop your pants... I think it is better to have behind the link

[example-safe-sync-unsafe-async]: #example-safe-sync-unsafe-async

### Async spawn
[example-async-spawn]: #example-async-spawn

Example from the ecosystem: [spawn_unchecked][spawn_unchecked-example-doc]

[spawn_unchecked-example-doc]: https://docs.rs/async-task/latest/async_task/fn.spawn_unchecked.html.

With the `Forget` trait we can make that API safe:

```rust
struct TaskHandler<'a>(u64, PhantomNonForget, PhantomData<&'a ()>);
// Or `struct TaskHandler<'a>(u64, PhantomNonForget<&'a ()>)`;

impl Drop for TaskHandler<'_> {
    fn drop(&mut self) {
        if let Some(mut mutex) = GLOBAL.get(self.0) {
            // We can block in async context as this mutex is held
            // during the `poll` which should return in a timely manner.
            let fut = mutex.lock();
            // cancel the future and call its drop handler
            drop(fut.take())
        }
    }
}

// Note that this is basically equivalent to async `scope`, as async `scope`
// would be transformed into the `Future` struct, just like `TaskHandler`.
fn spawn<'a>(fut: impl IntoFuture + 'a) -> TaskHandler<'a> {
    GLOBAL.spawn(fut)
}
```

### Async DMA
[example-async-dma]: #example-async-dma

> DMA stands for Direct Memory Access and it’s a peripheral used for transferring data between two memory locations in parallel to the operation of the core processor. For the purposes of this example, it can be thought of as `memcpy` in parallel to any other code.

Reviewer 6: DMA is used by various parts of the hardware, not just peripherals

```suggestion
DMA stands for Direct Memory Access, which is used to transfer data between two memory locations in parallel to the operation of the core processor. For the purposes of this example, it can be thought of as `memcpy` in parallel to any other code.
```

[uring-like design]: https://github.com/WebAssembly/component-model/issues/471

### `take_mut`

The async version of [`take_mut`] cannot be created as it relies on cleanup code to abort the program.

[`take_mut`]: https://docs.rs/take_mut/latest/take_mut/

### C/C++ bindings + async do not work well together
[example-async-c-cpp-bindings]: #example-async-c-cpp-bindings

It is common for C/C++ APIs to require some cleanup. It is not an issue for `sync` rust, as wrappers can just take a closure/callback and ensure that cleanup. But all `async` calls are transformed into `impl Future + use<'a>`, not passing control flow to the wrapper. `io_uring`, WASI 0.3 and `async-cuda` fall into that category too. By not having `!Forget` types Rust clearly lags behind current moments towards efficient asynchronous APIs. For embedded/kernel development this issue is even worse, as you often cannot afford an allocation due to the lack of resources or complex locking, making borrows your only option and making `Pin`'s drop guarantee not useful for you.

> # Guide-level explanation

Reviewer 6: There are a lot of detailed examples before the guide-level explanation, making this document hard to follow. It might help to choose one clear motivating example, and use that example in the motivation and guide sections.

Other examples could be moved to the reference-level explanation, or the migration or rationale sections - where this level of detail is more appropriate. (And more easily understood, after reading shorter motivation and guide-level explanations.)

Author: I fear that by reducing the motivation section the proposal may look less relevant, while it certainly is :sweat_smile:

Reviewer 6: If people don't understand the proposal within a few paragraphs, they will also struggle to see how it is relevant.

One way to do this is:
1. Describe one clear motivating example
2. Summarise or list the other motivating examples in one or two sentences
3. Say "for more motivating examples, see (later section)"

It might help to look at how other similar accepted RFCs have done it?

        let mut i = 0;
        loop {
            // `&mut` to `buf`.
            buf[i] = i;
            i = (i + 1) % 64;
            yield_now().await;
        }
    };

    let handle = spawn(fut);
    std::mem::forget(handle);

    // `fut` might still be running in the background, but `buf` is no longer protected by the borrow checker.

    // Undefined Behavior - aliasing a mutable reference.
    let fourth = buf[4];
}
```

In this case, `handle` borrows from `buf`, but the code that accessing `buf` is not directly tied to `handle`, it runs independently of it. Because of this, even if we pin `handle`, we still can *remove the borrow* (by ending the lifetime of `handle`) on `buf` while `JoinHandle`'s memory remains available (`forget(Box::pin(handle))`). Thus, `Pin` guarantees are not enough, we need `!Forget`.

Functions having signatures with weakening can remove a type from the scope without running its destructor. The following function is an example of a weakening function - after it is called, the borrow checker assumes that the lifetime of `T` has ended, as well as all borrows held by `T`.

```rust
fn weakener<T>(foo: T) -> i32 {
    std::mem::forget(T);
    0
}
```

> 

Reviewer 6: Would this section make more sense if located in the Rationale?
It doesn't seem to be required as part of the reference-level explanation or implementation.

Author: Rationale is about why this change over other alternatives, isn't it? This UB is what all alternative would try to solve, not what differentiates this proposal from others.

Reviewer 6: In the reference-level explanation, you can explain the undefined behaviour, and describe how it will be solved in detail.

In the rationale, you could explain the undefined behaviour that any alternative will need to resolve, and why this alternative is better at solving it than other alternatives.

    let arc1 = Arc::new(Mutex::new(None));
    let arc2 = arc1.clone();

    let mut buf = [0; 1024];
    let buf_ref = &mut buf;

    let handle = scoped(move || {
        let _handle = arc2.lock().unwrap().take();
        for _ in 0..100000 {
            black_box(&mut *buf_ref);
        }
        drop(arc2);
    });

    arc1.lock().unwrap().replace(handle);
    drop(arc1);

    // aliased `&mut`
    buf[0] = 1;
}
```

In this code, no memory is leaked, and `JoinHandle`'s destructor is not skipped. However, many types of channels - including rendezvous channels — can also be vulnerable to this issue if their signatures allow an equivalent implementation using reference counting.

```rust
fn main() {
    let (tx, rx) = std::sync::mpsc::channel();

    let mut buf = [0; 1024];
    let buf_ref = &mut buf;

    let handle = scoped(move || {
        let _handle = rx.recv().unwrap();
        for _ in 0..100000 {
            black_box(&mut *buf_ref);
        }
        drop(_handle);
    });

    tx.send(handle);
    drop(tx);

    buf[0] = 1;
}
```

### Solution for message passing of `!Forget` types.
[solution-to-self-referential-problem]: #solution-to-self-referential-problem

> One might speculate and try to fix some holes, for example by making `JoinHandle: !Send`, but this can only count as a workaround. If we look at the problem in depth, we can see that `Forget` is generally incompatible with `Rc`, as well as other APIs that can be expressed with its signature, because it creates a hidden self-reference. In the example earlier, the borrow checker cannot see a connection between `rx` and `tx` - when `tx` is dropped, `buf` is no longer borrowed. What if retained such a connection?

Reviewer 4: This reasoning is faulty as it cannot apply to rendezvous channels. Rendezvous channel buffer is simply a place on a waiting thread's stack.

Author: This paragraph is talking about lifetimes, and you still can have a lifetime connection between halves in rendezvous channel. Nothing is said about memory itself, as well as it's position.

Reviewer 4: > can have a lifetime connection between halves in rendezvous channel

You should substantiate why good old rendezvous channels either can or cannot operate on unforgettable objects.

Author: Without a connecting lifetime it allows types that are borrowed by `T` to become not borrowed while drop handler of `T` is not executed. This contradicts guarantee given by the reference-level explanation, thus those "good old rendezvous channels" (without a lifetime between `tx` and `rx`) are not allowed to work with `!Forget` types. To allow rendezvous channels work with `!Forget` types, `tx` and `rx` must be connected via lifetime and be `!Forget` themselves. Borrow checker already has all checks needed, but without a common lifetime it cannot ensure some properties, as discussed in the RFC.

struct Baz;

impl<T> Drop for Foo<T> { fn drop(&mut self) { } }

let ref_buf = [0u8; 64];
let mut_buf = [0u8; 64];

// We have a guarantee, that no `&mut` can be taken to `ref_buf` until `Foo`'s `drop`.
let foo_ref = Foo(PhantomNonForget, &ref_buf);
// We have a guarantee, that no `&/&mut` can be taken to `ref_buf` until `Foo`'s `drop`.
let foo_mut = Foo(PhantomNonForget, &mut ref_buf);

drop(ref_buf); // error[E0505]: cannot move out of `ref_buf` because it is borrowed
drop(mut_buf); // error[E0505]: cannot move out of `mut_buf` because it is borrowed

let ref_first_byte = ref_buf.0[0]; // Allowed
let mut_first_byte = mut_buf.0[0]; // error[E0503]: cannot use `mut_buf.0[_]` because it was mutably borrowed

drop(foo_ref);
drop(foo_mut);

let mut_first_byte = mut_buf.0[0]; // Allowed
drop(ref_buf); // Allowed
drop(mut_buf); // Allowed

// `Baz` cannot be moved or exclusively borrowed until `Foo` is dropped.
fn phantom<'a>(baz: &'a Baz) -> Foo<PhantomData<&'a ()>> {
    Foo(PhantomNonForget, PhantomData)
}
```

[`Pin::static_ref`]: https://doc.rust-lang.org/std/pin/struct.Pin.html#method.static_ref

Type becomes `!Forget` if it directly contains `!Forget` member.

> We should either allow `!Forget` types in statics or make all `'static` types `Forget` because it fulfills the unsafe guarantee and we can't enforce any code running before the program's abortion.

Reviewer 6: Standard exit is impacted too.

```suggestion
We should either allow `!Forget` types in statics or make all `'static` types `Forget` because it fulfills the unsafe guarantee and we can't enforce any code running before the program exits or aborts.
```

```rust
let mut resource = [0u8; 64];
let _unforget = Subsystem::execute(&mut resource);
std::process::abort(); // `resource` is (forcefully) borrowed for `'static`
resource[0] = 42; // unreachable
```

> ## Standard Library

Reviewer 6: Would this and following sections make more sense as part of the migration section?
It seems we start talking about migration here.

[std]: #std

All APIs in the standard library should be migrated at once. With available migration strategies, there is no benefit in gradual migration, it will greatly reduce the productivity of rustc developers by adding boilerplate and noise into the codebase. An audit must be performed to ensure which APIs must remain `Forget`. See [#migration](#migration) for more details.

> No existing non-generic types in std will be changed to `!Forget`.

Reviewer 6: Double negatives are confusing and harder for people to read and understand

```suggestion
All existing non-generic types in std will continue to be `Forget`.
```

// this `other_crate` and other crates can only observe `Trait::bar(..): Forget` cases.
mod other_crate {
    trait Trait {
        async fn baz();
    }
}
```

#### Macros

If macro-library generates code, some problems during the migration are possible:

```rust
mod user {
    #![default_generic_bounds(?Forget)]

    ::library::make!(); // Will not compile because `T` is `?Forget`.
}

mod user {
    #[macro_export]
    macro_rules! make {
        () => {
            pub fn foo<T>(t: T) {
                ::core::mem::forget(t);
            }
        }
    }
}
```

#### Changing default

It is not required, but in next editions we may swap the default for `default_generic_bounds`. Crates that want to continue using old default in next editions will set `#![default_generic_bounds(Forget)]`.

### Migration over the edition with [default auto traits]
[edition-migration-with-mask]: #edition-migration-with-mask

If [`local_default_bounds`] is not accepted, we can make a satisfactory migration by having editions <= 2024 have `Forget` as the default, and editions after 2024 have `?Forget` as the default.

> While it will not split the ecosystem, it will require everyone to make a migration just as in the [`local_default_bounds`] solution. There are concern about locking crates into `Forget` bounds on associated types in traits (like `async fn`). Migration can probably be automated for `#![forbid(unsafe)]` crates.

Reviewer 6: It's the absence of unsafe code that's the determining factor, not a specific lint against it, right?
(The migration tool could run the code for that lint before migrating itself, for example.)

```suggestion
While it will not split the ecosystem, it will require everyone to make a migration just as in the [`local_default_bounds`] solution. There are concern about locking crates into `Forget` bounds on associated types in traits (like `async fn`). Migration can probably be automated for crates without unsafe code.
```

We should either allow `!Forget` types in statics or make all `'static` types `Forget` because it fulfills the unsafe guarantee and we can't enforce any code running before the program's abortion.

```rust
let mut resource = [0u8; 64];
let _unforget = Subsystem::execute(&mut resource);
std::process::abort(); // `resource` is (forcefully) borrowed for `'static`
resource[0] = 42; // unreachable
```

## Standard Library
[std]: #std

All APIs in the standard library should be migrated at once. With available migration strategies, there is no benefit in gradual migration, it will greatly reduce the productivity of rustc developers by adding boilerplate and noise into the codebase. An audit must be performed to ensure which APIs must remain `Forget`. See [#migration](#migration) for more details.

No existing non-generic types in std will be changed to `!Forget`.

## `Copy`
[copy]: #copy

All types that implement `Copy` must implement `Forget` too.

## Unions
[unions]: #unions

Unions are always `Forget`. All members of `union` must be `Forget`, but it is already covered by other rules and does not need to be enforced.

## API changes
[library-api-changes]: #library-api-changes

> - `Rc`/`Arc` - all APIs for construction,  except the new `Rc::new_unchecked` method, only exist for `T: Forget` types. If we decide to not have `impl<T: 'static> Forget for T {}`, in the future we *may* allow safe constructors for `T: ?Forget + 'static` (resources are borrowed for `'static`, it fulfills the guarantee we are giving to the unsafe code) and something along the lines of `T: ?Forget + Freeze` (to forbid cycles), author of the RFC is not familiar enough with interior mutability questions.

Author: Something like a `Mutex` is also a problem, as can be seen in example from guide level explanation.

> - `Rc`/`Arc` - all APIs for construction,  except the new `Rc::new_unchecked` method, only exist for `T: Forget` types. If we decide to not have `impl<T: 'static> Forget for T {}`, in the future we *may* allow safe constructors for `T: ?Forget + 'static` (resources are borrowed for `'static`, it fulfills the guarantee we are giving to the unsafe code) and something along the lines of `T: ?Forget + Freeze` (to forbid cycles), author of the RFC is not familiar enough with interior mutability questions.

Author: I don't really think this is worth it to loosen anything at this point. We can always add a new constructor or loosen existing one. And the need for `Arc` in general is going to drop as you can just use what you are supposed to use, references.

> - `Rc`/`Arc` - all APIs for construction,  except the new `Rc::new_unchecked` method, only exist for `T: Forget` types. If we decide to not have `impl<T: 'static> Forget for T {}`, in the future we *may* allow safe constructors for `T: ?Forget + 'static` (resources are borrowed for `'static`, it fulfills the guarantee we are giving to the unsafe code) and something along the lines of `T: ?Forget + Freeze` (to forbid cycles), author of the RFC is not familiar enough with interior mutability questions.

Author: Thinking of this, there may be a third-party crate that will use `Rc::new_unchecked` to cover some cases where it is sound to have `?Forget` types.

- `ManuallyDrop<T>` always implements `Forget`, regardless of the `T`. `ManuallyDrop::new` is available for types with `T: Forget`.  New unsafe method `ManuallyDrop::new_unchecked`, available for `T: ?Forget`, is introduced. We may add a safe constructor with `T: ?Forget + 'static`, as we allow forgetting in statics.
- `Box::<T>::into_ptr` is available only for `T: Forget`. As for `T: !Forget` users should `ManuallyDrop::new_unchecked` and take the pointer via `&raw mut`. It will still be allowed to pass this pointer to `Box::from_ptr`.
- `Box::<T>::forget` is available only for `T: Forget`.
- `forget_unchecked`, a new unsafe function, is added to forget `T: ?Forget` types. It is a wrapper around `ManuallyDrop::new_unchecked`, just as `forget` is a wrapper around `ManuallyDrop::new`.
- `PhantomNonForget` is a `!Forget` ZST for types to become `!Forget`. If we decide to have `impl<T: 'static> Forget for T {}`, we should add a generic parameter/lifetime to `PhantomNonForget`.

> - `Vec::drain` is available only for `T: Forget` types. A new method might be added to work with `T: ?Forget`.

Reviewer 7: why not just mark the `Drain` iterator as `Forget` for `Vec<T>` only if `T: Forget`?

Author: This didn't cross my mind. I think it sounds good.
